### ISSUE TITLE
[Feature request] "Portable" settings (Windows)

### ISSUE THREAD

**Issue by vatterspun** [micahflee/onionshare#366]
Setup: Win7x86

**Current behavior:**
Saves settings to _USER\AppData\Roaming\OnionShare\onionshare.json_

**Requested behavior**

Settings written to the local folder (e.g. if the folder is at _C:\temp\OnionShare_, settings will be written to _C:\temp\OnionShare\onionshare.json_ ... if saved to _Z:\private\OnionShare_, it should write settings to _Z:\private\OnionShare\onionshare.json_)

**Why?**

Leaving files on the machine where the program is launched may give an indication that the program was used to transmit files, which could harm user privacy.  Additionally, if the program is run from a cloud service or USB drive, users will not need to re-enter settings on each computer the program is run from.

**Is this really an issue?**

I can't personally claim that a lot of users need this but I can say that Windows Tor Browser behaves this way, writing all settings to the application folder.

**Comment by mig5** [comment]
@micahflee I think this one was fixed in last release (the onionshare / onionshare-gui executables can take a --config argument to load settings from a user-specified location), maybe can be closed

**Comment by vatterspun** [comment]
Great news, thanks for the update.

**Comment by mig5** [comment]
Carrying from #551, not just the cache/state data but the config dir needs to be protected from other running OnionShare instances tampering with that data.

"But there's also another problem. Previously in Windows and Linux, you can open multiple instances of OnionShare, and each instance just had its own separate tor process and data dir. Now, if you open a second instance, it will crash and die because the first instance is using that data dir already. Maybe this is fine and you should only be allowed to run one instance at a time (or somehow multiple instances should share the same tor?), but if this is the case then we need make OnionShare check for other running instances when opening and handle it cleanly."

We need the config to be stored in a location that other PIDs can't write to (e.g the running PID claims some exclusive lock). We may as well store the Tor 'state' data in this place too for the same purpose.

**Comment by vatterspun** [comment]
> maybe the whole OnionShare installation code should be here too so that the whole thing could sit on a usb stick? (but not so sure about that)

Having all the files in one place as much as possible is ideal both for the purposes of running the program from a USB drive (and leave no traces behind) but also for saving inside an encrypted volume such as VeraCrypt.

Note I haven't tested the program's actual behavior yet (not for lack of desire, just circumstance) so I just wanted to state this as a goal rather than a bug.

**Comment by mig5** [comment]
I've heard in the field of more interest in this feature, particularly regarding Veracrypt volumes. We should step up this feature - adding to 2.1 milestone

**Comment by rysiekpl** [comment]
Yes, I can confirm this is an important improvement for investigative journalists and their sources in the field. It would make it possible to set up OnionShare inside a VeraCrypt container in a way that would provide plausible deniability to whomever is using it. This is a real-world scenario we have encountered.

**Comment by AdKiller** [comment]
Is portable setting implemented? Can it safely be sued on a USB drive?

**Comment by vatterspun** [comment]
Confirmed.  Writes to 

Great stuff, thanks.

**Comment by mig5** [comment]
@micahflee am I wrong or do we effectively have 'portable settings' now - as well as cache/state data - thanks to https://github.com/micahflee/onionshare/pull/1336 ? It might just be needed to be documented for advanced users storing their OnionShare data on a thumb drive or somesuch?

**Comment by snowbie** [comment]
> _EDIT_ I guess we do but only for Linux..
here's hoping for windows portability support
